### ISSUE TITLE
Spot removal tool

### ISSUE THREAD

**Issue by Beep6581** [Beep6581/RawTherapee#2239]
Originally reported on Google Code with ID 2255

```
I'm starting to code a dust/healing tool in the Pipette branch, but I'd like to know
if there's a difference in processing dust and healing?

I'll try to find the processing algorithm in Gimp's code base, but I'd appreciate if
someone could point me to this code directly, to save time.

Also, we have to find a place for this kind of tool. I'd suggest to create a new tab,
where this tool could be placed, as well as signature, framing. How should we name
this tab, and with which icon? I'd suggest "X-tra" as a name, and with a "pen" icon?
```

Reported by `natureh.510` on 2014-02-19 14:46:57

**Comment by Hombre57** [comment]
I've restarted the development of the Spot Removal tool. The GUI is functional, but I have some interrogations about how to deal with the pipeline.

In my current implementation, the processing is done on the image displayed in the Preview window **after** the transform processing. This means that each spot entry must have its source area completely visible and its destination area at least partly visible to process that entry. Also, if the zoom rate is too low (small image) and the final circle size is 4 pixels of diameter (IIRC), then that entry is not displayed either. So depending on the zoom rate and the position of the crop window, some Spot entry will be processed, and other won't.

You can also imagine other problem if the user change the transform parameters and the coarse transform ones.

The other way round would be to put this new tool before transformation (even coarse transform ?). In this case, when pressing on the Edit button, the initial image (corresponding to this step of the pipeline) would be displayed, and then the user will be able to put entries here and there in an absolute image space, whatever the transform will be right after. The downside of this is that I have to lock all other tools as long as we'll be in this Edit mode, otherwise the WB pipette or the rotate tool (e.g.) will not make much sens with this displayed image.

I can open a new branch if you need to dig in the code before answering, but I'd prefer to commit the enhanced-edit-mecanism first and rebase the Spot tool on the new master.

The other way round

**Comment by Hombre57** [comment]
Forgot to mention: if a source area is partly crossing the image border, the processing of that entry is skipped (but still exist as an entry).

**Comment by sguyader** [comment]
@Hombre57, thanks for trying to develop this tool. It is one of the most wanted tools left, and this will make RT the very best and complete raw development software.

**Comment by Beep6581** [comment]
> each spot entry must have its source area completely visible and its destination area at least partly visible to process that entry
It would be nice if the user was notified of this somehow, e.g. cursor changes, or the spot circle on the canvas gets a line through it or some kind of indication.

The LCP tool can lead to serious preview issues at certain zoom levels. This could be an argument for having spot healing before transform. Also remember that Ingo is working on lensfun support which could influence this tool.

Both solutions seem simple and intuitive, as long as the user gets feedback about what's happening.

**Comment by Hombre57** [comment]
In fact, it is even worse than that : let's say that entry 1 is not computed because out of scope, but entry 2 IS computed, but wrong because it's source area is the destination area of entry 1 (which should have been altered).

The modifications are re-entrant (as it should be), so ignoring some entry will make this tool quite tricky to use.

I think the best I should do is moving this tool before the Transform tool, where I'll be able to copy all required area from the InitialImage, even if out of scope. So, still some work to do :unamused:

**Comment by Hombre57** [comment]
I've looked at the transform function. It covers CA, Distortion, Rotation, Perspective, Gradient, PCVignetting, Vignetting, LCP.

Almost everything could be done after the Spot tool, but what about the CA feature (chromatic aberration) ? Wouldn't it produce better result for both the Spot tool and the CA tool to adjust CA before the Spot removal ? And is it okay to isolate CA from the lens distortion corrections ? @heckflosse @Beep6581 @michaelezra

**Comment by heckflosse** [comment]
@Hombre57 Which ca removal do you mean? We've three of them currently.

**Comment by heckflosse** [comment]
@Hombre57 Ca removal via LCP and/or via future LensFun support should not have a serious influence to Spot tool (if that's what you question)!

**Comment by Beep6581** [comment]
> Wouldn't it produce better result for both the Spot tool and the CA tool to adjust CA before the Spot removal

Yes, but.
1. Is it much more work for you to do that?
2. Will the spot tool be used on parts of the photo which are affected by CA? In other words, if the spot tool is used for dust removal from the sky, CA is not an issue there, it would only be an issue on non-plain areas. What can the spot tool do?
3. Is the course of action you take now limited in scope to this spot dust removal tool only, or will it influence any future local editing tools?

> is it okay to isolate CA from the lens distortion corrections ?

You would need to isolate:
1. Raw CA auto-correction,
2. RGB CA manual correction,
3. LCP CA correction,
4. and in the future lensfun CA correction.
Is that doable? And what does isolating them in the UI achieve?

**Comment by Hombre57** [comment]
@heckflosse
I'm referring to the `ImProcFunctions::needsTransform ()` function, which include the CA correction from `LensProfParams` and from `CACorrParams` (from the Raw tab).

@Beep6581 
1.1. it seems to be complex, at least for me, but I had to ask.
1.2. the raw tab's CA correction does shrink each r/g/b color channel differently, and thus change even plain area that generally has some grain
1.3. I don't really understand what you're saying here, but the I just want to insert the Spot removal tool to the right place in the pipeline. Actually, it groups some tools or filters into a single processing loop (e.g. with `ImProcFunctions::transformHighQuality` and `ImProcFunctions::rgbproc`), but if you want local editing, you'll have to split these group so you can apply multiple instance of their subpart (think about local editing of Exposure, Contrast, Saturation separatly). There's clearly an issue in the pipeline along the GUI problem.

> Would it make sense to the user?

Not really for the user, but for Image Quality.

> And what does isolating them in the UI achieve?

The problem is not in the UI, nothing has to be touched here, but in the pipeline.

**Comment by heckflosse** [comment]
@Hombre57 needsTransform does not include the ca correction params from Raw tab. And that's perfectly fine. Or do I miss anything?

**Comment by Hombre57** [comment]
@heckflosse You're right, it's refers to the Lens/Geometry CA corr. Sorry for the confusion.

**Comment by heckflosse** [comment]
@Hombre57 :)

**Comment by Hombre57** [comment]
Just for your information, I've uploaded my WIP patch of the Spot Removal tool. I'm afraid that I'll not be the one that will finish it, because of lack of time.

The GUI should be working, but the engine part is still to be done with the changes explained above.

If nobody take it over, it'll be a dead branch that you can delete if you want.

**Comment by Hombre57** [comment]
Sooooooooo... I had the time and will of merging dev into the spot-removal branch. If I find time to work on it again, I'd like to have answer to those questions first :

1. at which step should spot-removal be applied ?
2. what if the user transform the image (coarse transform, lens transform, rotate tool, ...) ? Do we want to update the point location ?

My thoughts on this is that it should be placed just after demozaicing and color-space conversion. The benefits are :

- you don't care of the transformations, the spots will always be removed because they are at fixed place. The image may be took in portrait mode, which mean that if you want to fix some dust on the sensor, we need to apply the same "initial" transform to the spot locations, i.e. the spot locations will have to be expressed in non-rotated and absolute coordinate (i.e. sensor coordinate)
- you don't need to process the source area, getting small parts of the buffered demozaiced image will be sufficient

The downside is that RT need to display that dull and untransformed image while editing the spots. If I manage to enter a special display for spot image, then it should be doable for the crop tool too. But that's a different story.

@agriggio @Beep6581 @Desmis @Floessie @heckflosse Do you agree with this behavior ? Any comment ?

**Comment by heckflosse** [comment]
@Hombre57 Maybe even before demosaic? But that's really just a thought...

**Comment by Desmis** [comment]
@Hombre57 
I think that "the sooner in the process" would be welcome, but it does not have to complicate the algorithm. Perhaps just after demoisaicing
jacques

**Comment by Beep6581** [comment]
It's a difficult question to answer not knowing the order of the pipeline. I would like to say that "just after demosaicing" is fine, but what about exposure adjustment or DCP tone curve? i.e. if I'm working on a HDR DNG which opens up as black and needs +6EV to see anything, will the spot preview be 0EV or +6EV? If 0EV then the user is "flying blind". Maybe that's ok in practice... I agree to go with what you suggested (just after demozaicing) and see how it works out.

**Comment by JGKle** [comment]
I was exploring options to move away from LR, and between DarkTable & RawTherapee, I do prefer RawTherapee.  However, spot removal is an essential tool that I use constantly.  Has this feature (no activity in over a year) been abandoned, or is it still somewhere on the horizon?

**Comment by Hombre57** [comment]
@metal450 It's not abandoned, I merged `dev` into `spot-removal` branch just yesterday, and am working on it actually. However RT's engine has not been designed for this kind of feature (especially the preview window), so it'll still require some time before it's complete. So yes it's on the horizon, but you know, the horizon is always at a constant distance from you :confused: , so I'll say it's on top of my TODO list.

**Comment by JGKle** [comment]
Ah - well, awesome to know it's at least still being worked on!  Thanks for your super quick reply :)

**Comment by Hombre57** [comment]
@metal450 It's here, at last ! Not merged yet, but very close to.

Please test the `spot-removal` branch. Shouldn't that tool be better named **Spot Healing** ?

## How to use the tool (i.e. documentation so far)

In the `Detail` tab, enable the `Spot Removal` tool and click on the _On preview editing_ button. The preview now display the image without crop and without processor intensive tools (e.g. Denoising, Wavelets, CIECam, Black & White, etc...), and most importantly, without transformation (Rotate, Perspective, Lens Correction [To be confirmed, I don't use it], etc).

To add a spot, press the `Ctrl` key while clicking over the location you want to edit, and drag the secondary circle to the source location. During the drag, you can release the `Ctrl` key.

By hovering the centers or the circles, you can move the source location, the destination location, the inner radius (where the effect is maximal) and the feather radius (where the effect progressively tends to zero). Those radius are paired, i.e. identical for the source and destination location.

To remove a spot, right click over it (this should probably change to Ctrl-Right click, for security reason ?). To remove all spots at once, click the `Reset` button in the tool's pane.

The tool's preview is visible at all zoom rates, as well as in detail windows, where the on preview widgets can be operated too. The preview is actually not shown in the thumbnails. When panning, the coarse image might not show the effect, or show it without color management. That's the last known caveat.

Some screenshots, for private use only (I don't have the permission from the model, though I cropped the image to make it anonymous).

[On preview editing](https://filebin.net/cfrbt3knevzf0mn2/SpotRemoval3.jpg?t=0sw8dgkq)
[Before/After view](https://filebin.net/cfrbt3knevzf0mn2/SpotRemoval2.jpg?t=0sw8dgkq)
[Output with tool disabled](https://filebin.net/cfrbt3knevzf0mn2/IMGP2193-0-without.jpg?t=0sw8dgkq)
[Output with tool enabled](https://filebin.net/cfrbt3knevzf0mn2/IMGP2193-1-with.jpg?t=0sw8dgkq)

**Comment by JGKle** [comment]
Very cool!!  Thanks for following up, & for the details :D

**Comment by Hombre57** [comment]
Note :To see the new icon from commit b4d7798, and if you've already built this branch, you'll have to cleanup the `cache/svg2png` folder, or at least delete the `spot-normal.png` file.

**Comment by Beep6581** [comment]
@Hombre57 shall we put this on the [5.7 milestone](https://github.com/Beep6581/RawTherapee/milestone/7) (out 4 days ago, woo!), or do you need a few months for shine and polish?

**Comment by heckflosse** [comment]
I played with the spot removal and healed some dust spots, which worked fine :+1: 
Then I selected neutral profile from profile combobox and got this crash:
```
(rawtherapee.exe:11060): glibmm-ERROR **: 11:42:45.659:
unhandled exception (type std::exception) in signal handler:
what: vector::_M_range_check: __n (which is 0) >= this->size() (which is 0)
```

I will try to reproduce...

**Comment by Hombre57** [comment]
@Beep6581 It'll require some polishing, I have another algorithm that I'd like to test first, instead of a simple cloning tool. If both are working, should we keep both and let the user select on a per Spot basis ? So yes, it'll be for 5.8.

**Comment by heckflosse** [comment]
Easy to reproduce:
1) create one healing spot
2) select neutral profile => crash

**Comment by heckflosse** [comment]
@Hombre57 I built using clang and got a lot of new warnings. See the attached log file
[clbuild.log](https://github.com/Beep6581/RawTherapee/files/3505253/clbuild.log)

**Comment by Hombre57** [comment]
I'm looking into the clang warning, @Floessie converted all `class` to `struct` in `procparams.h` in commit 02d6187cd880cd3b1a3804610e0e33ec181e9db0, and I'd love to know why, just out of curiosity. I hope to have solved everything, but I might need other `cbuild.log`. Could you make another try after pulling please ?

**Comment by heckflosse** [comment]
@Hombre57 Sure

**Comment by Hombre57** [comment]
It doesn't crash, but there's a problem with portrait shots, the spots are not always rendered in the preview (no problem for the output image I guess).

**Comment by heckflosse** [comment]
@Hombre57 

```
[ 44%] Building CXX object rtgui/CMakeFiles/rth.dir/batchqueuepanel.cc.obj
In file included from Z:/H2/rt56/rtgui/batchqueuepanel.cc:21:
In file included from Z:/H2/rt56/rtgui/preferences.h:26:
In file included from Z:/H2/rt56/rtgui/rtwindow.h:22:
In file included from Z:/H2/rt56/rtgui/filepanel.h:23:
In file included from Z:/H2/rt56/rtgui/batchtoolpanelcoord.h:23:
In file included from Z:/H2/rt56/rtgui/toolpanelcoord.h:55:
Z:/H2/rt56/rtgui/spot.h:106:17: warning: 'Spot::getCursor' hides overloaded virtual function [-Woverloaded-virtual]
    CursorShape getCursor (int objectID);
                ^
Z:/H2/rt56/rtgui/editcallbacks.h:74:25: note: hidden overloaded virtual function 'EditSubscriber::getCursor' declared here: different qualifiers ('const' vs unqualified)
    virtual CursorShape getCursor (int objectID) const;
                        ^
In file included from Z:/H2/rt56/rtgui/batchqueuepanel.cc:21:
In file included from Z:/H2/rt56/rtgui/preferences.h:26:
In file included from Z:/H2/rt56/rtgui/rtwindow.h:22:
In file included from Z:/H2/rt56/rtgui/filepanel.h:23:
In file included from Z:/H2/rt56/rtgui/batchtoolpanelcoord.h:23:
In file included from Z:/H2/rt56/rtgui/toolpanelcoord.h:55:
Z:/H2/rt56/rtgui/spot.h:120:10: warning: 'tweakParams' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
    void tweakParams(rtengine::procparams::ProcParams& pparams);
         ^
Z:/H2/rt56/rtgui/../rtengine/tweakoperator.h:42:18: note: overridden virtual function is here
    virtual void tweakParams(procparams::ProcParams& pparams) = 0;
                 ^
2 warnings generated.
[ 45%] Building CXX object rtgui/CMakeFiles/rth.dir/batchtoolpanelcoord.cc.obj
In file included from Z:/H2/rt56/rtgui/batchtoolpanelcoord.cc:20:
In file included from Z:/H2/rt56/rtgui/batchtoolpanelcoord.h:23:
In file included from Z:/H2/rt56/rtgui/toolpanelcoord.h:55:
Z:/H2/rt56/rtgui/spot.h:106:17: warning: 'Spot::getCursor' hides overloaded virtual function [-Woverloaded-virtual]
    CursorShape getCursor (int objectID);
                ^
Z:/H2/rt56/rtgui/editcallbacks.h:74:25: note: hidden overloaded virtual function 'EditSubscriber::getCursor' declared here: different qualifiers ('const' vs unqualified)
    virtual CursorShape getCursor (int objectID) const;
                        ^
In file included from Z:/H2/rt56/rtgui/batchtoolpanelcoord.cc:20:
In file included from Z:/H2/rt56/rtgui/batchtoolpanelcoord.h:23:
In file included from Z:/H2/rt56/rtgui/toolpanelcoord.h:55:
Z:/H2/rt56/rtgui/spot.h:120:10: warning: 'tweakParams' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
    void tweakParams(rtengine::procparams::ProcParams& pparams);
         ^
Z:/H2/rt56/rtgui/../rtengine/tweakoperator.h:42:18: note: overridden virtual function is here
    virtual void tweakParams(procparams::ProcParams& pparams) = 0;
                 ^
2 warnings generated.
```

**Comment by Hombre57** [comment]
@heckflosse Could you give a last try ? I should have removed all the warnings now.  The bug I reported in my last comment is solved too.

**Comment by heckflosse** [comment]
@Hombre57 Sure, but will take a bit longer this time as I'm just compiling capture_sharpening branch...

**Comment by heckflosse** [comment]
@Hombre57 Compiling latest spot_removal now...

**Comment by heckflosse** [comment]
@Hombre57 

```
Z:/H2/rt56/rtgui/spot.h:106:17: warning: 'getCursor' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
    CursorShape getCursor (int objectID) const;
                ^
Z:/H2/rt56/rtgui/editcallbacks.h:74:25: note: overridden virtual function is here
    virtual CursorShape getCursor (int objectID) const;
                        ^
1 warning generated.
```

**Comment by Hombre57** [comment]
@heckflosse It will take several centuries since I'll get used to this `override` thing...

**Comment by Hombre57** [comment]
@heckflosse Done in latest commit, hopefully.

**Comment by heckflosse** [comment]
@Hombre57 Fixes confirmed

**Comment by Floessie** [comment]
Builds and runs fine with GCC 8.3 on Debian 10 amd64/i386, Clang 7 on Debian 10 amd64, and GCC 9.2 as well as Clang 8 on Debian unstable ppc.

The UI is super intuitive, slim and easy to use. :+100:

> I'm looking into the clang warning, @Floessie converted all class to struct in procparams.h in commit 02d6187, and I'd love to know why, just out of curiosity.

A `struct` usually indicates the use as a data-only, freely accessable, OO-agnostic collection of elements, whereas `class` is used for more sophisticated things (e.g. a `class` has getters and setters to access members). I classified the ProcParams data collections as mainly data-only, thus I chose `struct` for them. Sorry, if that caused you trouble.

Best,
Flössie

@heckflosse Here's a patch to fix some Clang 8 warnings on non-SSE2 builds:

```patch
diff --git a/rtengine/amaze_demosaic_RT.cc b/rtengine/amaze_demosaic_RT.cc
index ffb68dbdd..55e6f7281 100644
--- a/rtengine/amaze_demosaic_RT.cc
+++ b/rtengine/amaze_demosaic_RT.cc
@@ -768,7 +768,7 @@ void RawImageSource::amaze_demosaic_RT(int winx, int winy, int winw, int winh, c
 
                         //if both agree on interpolation direction, choose the one with strongest directional discrimination;
                         //otherwise, choose the u/d and l/r difference fluctuation weights
-                        if ((0.5 - varwt) * (0.5 - diffwt) > 0 && fabsf(0.5 - diffwt) < fabsf(0.5 - varwt)) {
+                        if ((0.5f - varwt) * (0.5f - diffwt) > 0.f && fabsf(0.5f - diffwt) < fabsf(0.5f - varwt)) {
                             hvwt[indx >> 1] = varwt;
                         } else {
                             hvwt[indx >> 1] = diffwt;
@@ -1236,7 +1236,7 @@ void RawImageSource::amaze_demosaic_RT(int winx, int winy, int winw, int winh, c
                         //first ask if one gets more directional discrimination from nearby B/R sites
                         float pmwtalt = xdivf(pmwt[(indx - m1) >> 1] + pmwt[(indx + p1) >> 1] + pmwt[(indx - p1) >> 1] + pmwt[(indx + m1) >> 1], 2);
 
-                        if (fabsf(0.5 - pmwt[indx1]) < fabsf(0.5 - pmwtalt)) {
+                        if (fabsf(0.5f - pmwt[indx1]) < fabsf(0.5f - pmwtalt)) {
                             pmwt[indx1] = pmwtalt;   //a better result was obtained from the neighbours
                         }
 
@@ -1304,7 +1304,7 @@ void RawImageSource::amaze_demosaic_RT(int winx, int winy, int winw, int winh, c
 
                     for (int cc = 12 + (FC(rr, 2) & 1), indx = rr * ts + cc, indx1 = indx >> 1; cc < cc1 - 12; cc += 2, indx += 2, indx1++) {
 
-                        if (fabsf(0.5 - pmwt[indx >> 1]) < fabsf(0.5 - hvwt[indx >> 1]) ) {
+                        if (fabsf(0.5f - pmwt[indx >> 1]) < fabsf(0.5f - hvwt[indx >> 1]) ) {
                             continue;
                         }
 
```

**Comment by Hombre57** [comment]
Thanks for the explanation @Floessie, it didn't caused trouble, but I didn't knew the difference and use case. I'll convert the `Rectangle` class to a struct on next commit then.

**Comment by Floessie** [comment]
@Hombre57 You mean the one in `editwidgets.h`?

**Comment by Floessie** [comment]
When editing the spots in similar regions, it's not immediately clear, which of the linked spots is the source and which one is the destination. Would it be possible to have a little indicator like an arrow head towards the destination?

**Comment by Thanatomanic** [comment]
@Hombre57 I just tested your tool and it works really well! Well done. 
A few things:
- ~~Please indicate which spot is the source and which is the destination~~ @Floessie was first
- Is there a techical reason why the feather is rather limited? Is it reasonable to make it at least as big as the chosen spot diameter? 
- Working on spots in 1:1 zoom is relatively slow. Not so surprising of course, when you consider the image needs full processing with every change. Still, not ideal.
- Would it be possible to make the feather line thinner or dotted/dashed? This gives a better visual cue to its purpose imo, and is seen in other software too.
- When you enable the tool, this icon is different from the tool's icon. Not sure if that's intended, or desirable.
![image](https://user-images.githubusercontent.com/6567747/63159544-17a7c900-c01c-11e9-9b0b-7868a3fcf7f8.png)
- A little bug: enable the tool, then F3/F4 to another image. The icon still shows the tool as active, but it isn't.

**Comment by Hombre57** [comment]
@Floessie 

> You mean the one in editwidgets.h?

No I meant from rtengine/spot.h

> When editing the spots in similar regions, it's not immediately clear, which of the linked spots is the source and which one is the destination. Would it be possible to have a little indicator like an arrow head towards the destination?

Yeah, I'm also annoyed by not distinguishing source and dest. I'll think about something, but I don't think it can be on the linking line, as is can be very short or non existent.

@Thanatomanic 

> Is there a techical reason why the feather is rather limited? Is it reasonable to make it at least as big as the chosen spot diameter?

I don't think it's so limited : if inner spot's radius is R, the feather max radius is 2xR. Isn't this enough ? Would like the opinion from others too.

> Working on spots in 1:1 zoom is relatively slow. Not so surprising of course, when you consider the image needs full processing with every change. Still, not ideal.

It's quite fast here, of course for Release builds and with latest commit. When you click on the Edit button, it enter into a mode where some tools are deactivated, because they are too time consuming or change the geometry (so all tools from the Geometry tab are disabled), excepted the Coarse transform (the image is already _coarse transformed_ at the location of the Spot tool in the pipeline). I still kept the Exposure tool active, as well as Tone Compression and @agriggio 's HDR Tone compression, so that you won't have dark HDR images when editing the spots. The full image is not entirely processed for this tool, however it's almost at the beginning of the pipeline.

If I forgot to disable some tools, you can have a look [here ](https://github.com/Beep6581/RawTherapee/blob/0d19cd40a6371bbcde6b9939149ef06cbca4880f/rtgui/spot.cc#L729) to tweak the procparams during the spot editing, and tell what is missing.

> Would it be possible to make the feather line thinner or dotted/dashed? This gives a better visual cue to its purpose imo, and is seen in other software too.

Ok, I'll do that.

> When you enable the tool, this icon is different from the tool's icon. Not sure if that's intended, or desirable.

The icon shown in the toolbar is common to all on preview editing, would it be curve adjustment or operation on widgets. Probably we should find a more general icon, but I'm against having one more icon.

> A little bug: enable the tool, then F3/F4 to another image. The icon still shows the tool as active, but it isn't.

I'll have a look.

Thanks for the testings !

**Comment by Hombre57** [comment]
@Thanatomanic @Beep6581 

> A little bug: enable the tool, then F3/F4 to another image. The icon still shows the tool as active, but it isn't.

Fixed in branch `editbuttonfix` because bug in `dev` too. Could you test ? It's a minor fix and should be part of 5.7 IMHO.

**Comment by Beep6581** [comment]
> I'm also annoyed by not distinguishing source and dest. I'll think about something, but I don't think it can be on the linking line, as is can be very short or non existent.

You could make one of the circle's outlines a different thickness or style to the other, e.g. dotted.

**Comment by Desmis** [comment]
@Hombre57 

I come back at home... with Internet !
I am always in vacation with my family, but I test this branch

Works fine, idea evoqued by @Beep6581  is good  :)

jacques

**Comment by Hombre57** [comment]
Branch `editbuttonfix` merged to dev (commit 6b12f299e698be8e26efd30f8594afb56bf00d6a ), which has been merged back in `spot-removal-tool`.

**Comment by Floessie** [comment]
@Hombre57 
> No I meant from rtengine/spot.h

Then: yes. :+1: 

Here's a little compaction patch (only test-compiled):

```patch
diff --git a/rtengine/spot.cc b/rtengine/spot.cc
index 2244d882e..dc8f92c70 100644
--- a/rtengine/spot.cc
+++ b/rtengine/spot.cc
@@ -23,18 +23,6 @@
 #include "imagesource.h"
 #include <iostream>
 
-namespace
-{
-
-// "ceil" rounding
-template<typename T>
-constexpr T skips(T a, T b)
-{
-    return a / b + static_cast<bool>(a % b);
-}
-
-}
-
 namespace rtengine
 {
 
@@ -48,14 +36,12 @@ public:
     };
 
     struct Rectangle {
-    public:
         int x1;
         int y1;
         int x2;
         int y2;
 
-        Rectangle() : x1(0), y1(0), x2(0), y2(0) {}
-        Rectangle(const Rectangle &other) : x1(other.x1), y1(other.y1), x2(other.x2), y2(other.y2) {}
+        Rectangle() : Rectangle(0, 0, 0, 0) {}
         Rectangle(int X1, int Y1, int X2, int Y2) : x1(X1), y1(Y1), x2(X2), y2(Y2) {}
 
         bool intersects(const Rectangle &other) const {
@@ -65,23 +51,23 @@ public:
 
         bool getIntersection(const Rectangle &other, std::unique_ptr<Rectangle> &intersection) const {
             if (intersects(other)) {
-                if (!intersection) {
-                    intersection.reset(new Rectangle());
-                }
-                intersection->x1 = rtengine::max(x1, other.x1);
-                intersection->x2 = rtengine::min(x2, other.x2);
-                intersection->y1 = rtengine::max(y1, other.y1);
-                intersection->y2 = rtengine::min(y2, other.y2);
-
-                if (intersection->x1 > intersection->x2 || intersection->y1 > intersection->y2) {
-                    intersection.release();
+                std::unique_ptr<Rectangle> intsec(
+                    new Rectangle(
+                        rtengine::max(x1, other.x1),
+                        rtengine::max(y1, other.y1),
+                        rtengine::min(x2, other.x2),
+                        rtengine::min(y2, other.y2)
+                    )
+                );
+
+                if (intsec->x1 > intsec->x2 || intsec->y1 > intsec->y2) {
                     return false;
                 }
+
+                intersection = std::move(intsec);
                 return true;
             }
-            if (intersection) {
-                intersection.release();
-            }
+
             return false;
         }
 
@@ -101,15 +87,15 @@ public:
             return *this;
         }
 
-        Rectangle& operator/=(const int &v) {
+        Rectangle& operator/=(int v) {
             if (v == 1) {
                 return *this;
             }
 
             int w = x2 - x1 + 1;
             int h = y2 - y1 + 1;
-            w = w / v + (w % v > 0);
-            h = h / v + (h % v > 0);
+            w = w / v + static_cast<bool>(w % v);
+            h = h / v + static_cast<bool>(h % v);
             x1 /= v;
             y1 /= v;
             x2 = x1 + w - 1;
@@ -117,7 +103,7 @@ public:
 
             return *this;
         }
-};
+    };
 
 private:
     Type type;
@@ -173,7 +159,7 @@ public:
         }
     }
 
-    SpotBox& operator /=(const int& v) {
+    SpotBox& operator /=(int v) {
         if (v == 1) {
             return *this;
         }
```

> Yeah, I'm also annoyed by not distinguishing source and dest. I'll think about something, but I don't think it can be on the linking line, as is can be very short or non existent.

How about a litte "S" and "D" in a corner of the spot's bounding box?

**Comment by Beep6581** [comment]
S and D are not localizable.

**Comment by Floessie** [comment]
Do they have to? Any alternative unambiguous symbol? Coloring the circles?

**Comment by Hombre57** [comment]
@Floessie Thanks for the patch, I'll add it to the next commit.

To differentiate S and D, I'll opt for dashed line for the source inner circle.

I'm actually trying to solve issue #4858 for 5.7, and possibly #5310 which seem connected, so the Spot tool will wait few days.

**Comment by Beep6581** [comment]
@Floessie well yes, RT should be localizable. I suggested dotted vs solid lines, but other software, such as the one from that Mexican housing company, use yellow vs white circles.

**Comment by Hombre57** [comment]
@Floessie @Beep6581 I made some try with dashed circles for source location : for small circles it's very ugly.

I think differentiating by the color would be a simpler and more elegant solution. I'd suggest :

1. source : white / dest : green
2. source : white / dest : yellow
3. source : yellow / dest : green
4. ???

Circles are orange when hovered, and red when resized. The link will stay white.

**Comment by Beep6581** [comment]
Better to avoid strong colors over the preview (same reason RT uses a neutral color scheme by default), so I vote for 2.

**Comment by Beep6581** [comment]
I was wrong about the Mexican brick manufacturer (the yellow color was actually from the screencasting software), it uses varying line thickness and an arrow - https://youtu.be/vi9JKMIb7Vw?t=106

**Comment by Hombre57** [comment]
@Beep6581 & all : I've added source/dest differentiation with dashed circles. If you find this ugly, I can switch to plain circles and yellow destination.

**Comment by Hombre57** [comment]
@Beep6581 This patch won't be ready for 5.8 (unless I can find some free time and you're okay with that), but contrary to my initial plan, I won't be able to add a real spot healing tool. The actual feature is _only_ a clone tool that can serve to suppress unwanted spots.

I don't think it's worth to postpone its merge into dev because of that. It'll still be possible to add a new mode later one, with graphical element to switch between them.

**Comment by Beep6581** [comment]
@Hombre57 then let's put it on the 5.9 milestone.

**Comment by eathtespagheti** [comment]
can't wait to see this in 5.9!!

**Comment by Hombre57** [comment]
@eathtespagheti The Spot tool is somewhat ready, the only big remaining problem is _coarse transformation invariance_. I mean that if you rotate or flip the image after the placement of the spots, they won't follow the transformation.

That's something I wished to avoid, but `ImageSource::getImage` include the coarse transformation, and that's where I'm actually stuck.

So if you, @Beep6581, @heckflosse and others agree on committing as is, it can be done rapidly.

**Comment by Lawrence37** [comment]
I see two branches: spot-removal-tool and spotremov. spotremov has one more commit, a merge from dev. What is the purpose of having the two branches?

**Comment by Desmis** [comment]
@Lawrence37 
I open this branch one mounth ago...

Why ?
I you look to ART, you will see that "Spot removal" has been improved by the addition of several sliders to better manage transitions and opacity.

I tried to carry these modifications, but in front of the difficulties (GUI, code...) I gave up.

Now, what to do? Leave "Spot removal tool"  as it is or try to bring the modifications of ART...maybe for 6.0...or latter ?

Jacques

**Comment by Thanatomanic** [comment]
I believe the `spot-removal-tool` is the branch Hombre worked on. I believe it is working can could easily be converted into a PR for testing. There are of course tweaks we could implement and maybe bring over from ART, but even rudimentary spot removal is better than none imo.

**Comment by Beep6581** [comment]
Merged for 5.9. Needs to be documented - #6247
